dashboard/app/helpers/aichat_openai_helper.rb

```diff
@@ -72,6 +76,60 @@ def self.get_instructions(system_prompt, level_system_prompt, retrieval_contexts
     instructions
   end
 
+  # Reports and logs usage metrics to Cloudwatch
+  def self.report_usage_metrics(usage, messages, level_id, encrypted_channel_id, user_id)
+    return unless usage
+
+    filtered = messages.reject {|message| message[:content].is_a?(String)}
+    messages_with_assets_count = filtered.count {|message| message[:content].any? {|c| c[:type] != 'text'}}
+    pdfs_count = filtered.sum {|message| message[:content].count {|c| c[:type] == 'file'}}
+    images_count = filtered.sum {|message| message[:content].count {|c| c[:type] == 'image_url'}}
+
+    is_multimodal = messages_with_assets_count > 0
+
+    input_rate = 0.15 / 1_000_000 # $0.15 per million tokens
+    output_rate = 0.60 / 1_000_000 # $0.60 per million tokens
+    input_cost = usage['prompt_tokens'] * input_rate
+    output_cost = usage['completion_tokens'] * output_rate
+    total_cost = input_cost + output_cost
+
+    log_payload = {
+      event: 'aichat_openai_usage',
+      multimodal: is_multimodal,
+      usage: usage,
+      messages: {
+        total: messages.count,
+        withAssets: messages_with_assets_count,
+        pdfs: pdfs_count,
+        images: images_count,
+      },
+      cost: {
+        input: "$#{format("%.6f", input_cost)}",
+        output: "$#{format("%.6f", output_cost)}",
+        total: "$#{format("%.6f", total_cost)}"
+      },
+      levelId: level_id,
+      channelId: encrypted_channel_id,
+      userId: user_id
+    }
+
+    CDO.log.info log_payload.to_json.to_s if DCDO.get('log_aichat_openai_usage', true)
+
+    metrics = ['prompt_tokens', 'completion_tokens'].map do |key|
+      {
+        metric_name: "AichatOpenaiRequest.#{key.camelize(:upper)}",
+        value: usage[key],
```

Comment:
Thoughts on whether to have the metric being the raw # of tokens vs. actual $$ spent?

Reply:
Good question, I figured technically the token amount is more precise and we can multiply by cost in the graph directly? That way we're not also super locked in case the rate were to change for any reason. I did include cost in the log payload just out of visual convenience but we have the tokens there too in case we need a more precise calculation.